libbpf-tools/killsnoop.c

```diff
@@ -0,0 +1,180 @@
+            break;
+        case 'h':
+            argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+            break;
+        default:
+            return ARGP_ERR_UNKNOWN;
+    }
+    return 0;
+}
+
+static void sig_int(int signo)
+{
+    exiting = 1;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+    const struct event *e = data;
+
+    /*
+       double ts = 0.0;
+       static __u64 start_timestamp = 0;
+       int fd, err;
+       if (e->ret >= 0) {
+       fd = e->ret;
+       err = 0;
+       } else {
+       fd = -1;
+       err = -e->ret;
+       }
+       if (!start_timestamp)
+       start_timestamp = e->ts_ns;
+       if (emit_timestamp) {
+       ts = (double)(e->ts_ns - start_timestamp) / 1000000000;
+       printf("%-14.9f ", ts);
+       }
+       */
+    printf("%-7d %-16s %-9d %-8d %d\n", 
+            e->pid, e->comm, (e->sig), e->tpid, e->ret);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+    warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+int main(int argc, char **argv)
+{
+    static const struct argp argp = {
+        .options = opts,
+        .parser = parse_arg,
+        .doc = argp_program_doc,
+    };
+    struct perf_buffer_opts pb_opts;
+    struct perf_buffer *pb = NULL;
+    struct killsnoop_bpf *obj;
+    int err;
+
+    err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+    if (err)
+        return err;
+
+    err = bump_memlock_rlimit();
+    if (err) {
+        warn("failed to increase rlimit: %d\n", err);
+        return 1;
+    }
+
+    obj = killsnoop_bpf__open();
+    if (!obj) {
+        warn("failed to open BPF object\n");
+        return 1;
+    }
+
+    obj->rodata->target_pid = target_pid;
+    obj->rodata->trace_failed_only = trace_failed_only;
+
+    err = killsnoop_bpf__load(obj);
+    if (err) {
+        warn("failed to load BPF object: %d\n", err);
+        goto cleanup;
+    }
+
+    err = killsnoop_bpf__attach(obj);
+    if (err) {
+        warn("failed to attach BPF programs: %d\n", err);
+        goto cleanup;
+    }
+    pb_opts.sample_cb = handle_event;
+    pb_opts.lost_cb = handle_lost_events;
+    pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES,
+            &pb_opts);
+    err = libbpf_get_error(pb);
+    if (err) {
+        warn("failed to open perf buffer: %d\n", err);
+        goto cleanup;
+    }
+
+    if (signal(SIGINT, sig_int) == SIG_ERR) {
+        warn("can't set signal handler: %s\n", strerror(-errno));
+        goto cleanup;
+    }
+
+    if (emit_timestamp)
+        printf("%-14s ", "TIME(s)");
+    printf("%-7s %-16s %-9s %-8s %s\n", 
+            "PID", "COMM", "SIG", "TPID", "RESULT");
+
+    while (1) {
+        if ((err = perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS)) < 0)
+            break;
+        if (exiting)
+            goto cleanup;
+    }
+    warn("error polling perf buffer: %d\n", err);
+cleanup:
+    killsnoop_bpf__destroy(obj);
+
```

Comment:
Could you please fix follow the same approach as in https://github.com/iovisor/bcc/pull/3630?